### PR TITLE
107 users can scroll through and interact with storylines

### PR DESCRIPTION
**Pull request by broekhuisg**
@antw Hi. I created this pull request, but it's not fully finished yet.
Could you maybe help me with some typescript issues? If that is fixed, then the pull request can be completed if you ask me.

In the file Section.tsx I have a type with a variable called 'value', which can me multiple types. But then vscode is complaining about that is not correct. He keeps saying that 'value is not in Slider | StaticImage | string'. Can you help me out on this one?

Thank you in advance

### CODE REVIEW

**Comment by antw** [comment]
> In the file Section.tsx I have a type with a variable called 'value', which can me multiple types. But then vscode is complaining about that is not correct. He keeps saying that 'value is not in Slider | StaticImage | string'. Can you help me out on this one?

This is because [Content.value](https://github.com/ZEnMo/Holon-webapp/blob/047161c8ae9a69f28208dd34b4918f58f746a54f/frontend/components/Section/Section.tsx#L9) is a union type. It says that the value may be a `Slider` **or a** `StaticImage` **or a** `string`. You have to handle each one of those possibilities. You are doing that by switching on `content.type`, but TS isn't smart enough to know what `value` is.

Since `value` is dependent on `type`, we can do this:

```ts
export type Content =
  | {
      id?: string;
      type: "text";
      value: string;
    }
  | {
      id?: string;
      type: "slider";
      value: Slider;
    }
  | {
      id?: string;
      type: "static_image";
      value: StaticImage;
    };
```

TypeScript now knows that`value` depends on `type`. For example, when `type` is "slider" the `value` *must* be a `Slider`. With this change in place, the switch statements work as expected.

Here's a full diff:

```diff
diff --git a/frontend/components/Section/Section.tsx b/frontend/components/Section/Section.tsx
index 1bc069c..6c28667 100644
--- a/frontend/components/Section/Section.tsx
+++ b/frontend/components/Section/Section.tsx
@@ -3,11 +3,22 @@ import ImageSlider from "@/components/InteractiveImage/ImageSlider";
 import Image from "next/image";
 import RawHtml from "@/components/RawHtml/RawHtml";

-export type Content = {
-  id?: string;
-  type?: string;
-  value?: Slider | StaticImage | string;
-};
+export type Content =
+  | {
+      id?: string;
+      type: "text";
+      value: string;
+    }
+  | {
+      id?: string;
+      type: "slider";
+      value: Slider;
+    }
+  | {
+      id?: string;
+      type: "static_image";
+      value: StaticImage;
+    };

 export type Slider = {
   id?: number;
@@ -41,13 +52,12 @@ export default function Section({ data }) {
         case "slider":
         case "text":
           if (content.type == "slider") {
-            content!.value.currentValue = content.value.sliderValueDefault;
+            content.value.currentValue = content.value.sliderValueDefault;
           }
           contentArr.push(content);
           break;
         case "static_image":
-          console.log(content);
-          setMedia(content);
+          setMedia(content.value);
           break;
         default:
       }
@@ -58,7 +68,7 @@ export default function Section({ data }) {

   const setSliderValue = id => {
     const currentSlider = content.find(slider => slider.id == id);
-    if (currentSlider !== undefined) {
+    if (currentSlider !== undefined && currentSlider.type === "slider") {
       currentSlider.value.currentValue = value;
       const currentSliderIndex = content.findIndex(slider => slider.id == id);
       const spreadedSliders = [...content];
@@ -77,7 +87,7 @@ export default function Section({ data }) {
     <div className="storyline__row flex flex-col lg:flex-row">
       <div className="flex flex-col p-8 lg:w-1/3 bg-slate-200">
         {content.map((ct, _index) => {
-          if (ct.type == "slider") {
+          if (ct.type === "slider") {
             return (
               <ImageSlider
                 key={`slider${_index}`}
```

There is still an issue further down [when creating an `ImageSlider`](https://github.com/ZEnMo/Holon-webapp/blob/047161c8ae9a69f28208dd34b4918f58f746a54f/frontend/components/Section/Section.tsx#L82-L94). As Content.value may be `undefined`, some of those props are not valid since they require a concrete value. For example, ImageSlider.value **must** be a number, but `ct.value?.currentValue` is `number | undefined`.

File: frontend/components/Section/Section.tsx [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        case "text":
+          if (content.type == "slider") {
+            content!.value.currentValue = content.value.sliderValueDefault;
+          }
+          contentArr.push(content);
+          break;
+        case "static_image":
+          console.log(content);
+          setMedia(content);
+          break;
```
Comment (mattijsstam) [inline frontend/components/Section/Section.tsx:51]: There is no case for animation in this version is this addition documented in another story or somewhere else?

File: frontend/components/Storyline/Storyline.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        <p>
+          <RawHtml html={exampledata.description} />
+        </p>
+      </div>
+      {exampledata.body.map((scenario, index) => (
+        <StorylineScenario key={index} storylineScenario={scenario} />
+      ))}
+    </>
+  );
+}
```
Comment (mattijsstam) [inline frontend/components/Storyline/Storyline.tsx:21]: This file is only referenced on the old homepage but gave me quite some confusion during the review. I think it's better to remove all references to the old storylines from the code.

File: src/main/pages/scenario.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+from django.db import models
```
Comment (mattijsstam) [inline src/main/pages/scenario.py:1]: I'm not sure why this model and related pages are in this branch and how to relate this to the design

**Review by mattijsstam — CHANGES_REQUESTED** [review]
Overall it works but there are still some loose ends in this branch

- [ ] There is an aditional model scenario which doesn't work in the frontend and can't be related to any story or design (by me)
- [ ] The design of the storyline page deviates quite a bit from the UX design (no coloured header, different headers and fonts, other slider design
- [ ] Images in the storyline page are streched and not cropped using the image renditions in wagtail https://docs.wagtail.org/en/v4.1/topics/images.html

I think at least the 1st and 3rd point should be looked at before merging. The design part can also be a new story but it has to land somewhere.

**Comment by mattijsstam** [comment]
When uploading a video the following error is thrown in the backend

```
Request Method: | GET
-- | --
http://localhost:3000/wt/api/nextjs/v1/page_by_path/?html_path=storyline-overview%2Fstoryline
4.1.2
AttributeError
'EmbedValue' object has no attribute 'file'
/workspace/src/main/blocks/text_and_media.py, line 41, in get_api_representation
nextjs.api.PageByPathAPIViewSet
/usr/local/bin/python
3.11.0
['/workspace/src',  '/workspace/src',  '/workspace',  '/usr/local/lib/python311.zip',  '/usr/local/lib/python3.11',  '/usr/local/lib/python3.11/lib-dynload',  '/usr/local/lib/python3.11/site-packages']
Fri, 04 Nov 2022 08:25:58 +0100
```

**Comment by mattijsstam** [comment]
@pascalalferink de content in een storyline heeft nog geen styling, ik heb geprobeerd de storyline van het design na te maken maar het resultaat is nog niet helemaal volgens verwachting.

image

Het zou mooi zijn als het er zoveel mogelijk uitziet als het [design](https://xd.adobe.com/view/fdfd0445-f79f-426d-a864-c0cb70e3329b-a8ae/screen/2b23d2b5-55ff-4888-a092-b86df8167083/specs/) dit is wel wat de mensen aankomende dinsdag verwachten.

**Comment by broekhuisg** [comment]
> When uploading a video the following error is thrown in the backend
> 
> ```
> Request Method: | GET
> -- | --
> http://localhost:3000/wt/api/nextjs/v1/page_by_path/?html_path=storyline-overview%2Fstoryline
> 4.1.2
> AttributeError
> 'EmbedValue' object has no attribute 'file'
> /workspace/src/main/blocks/text_and_media.py, line 41, in get_api_representation
> nextjs.api.PageByPathAPIViewSet
> /usr/local/bin/python
> 3.11.0
> ['/workspace/src',  '/workspace/src',  '/workspace',  '/usr/local/lib/python311.zip',  '/usr/local/lib/python3.11',  '/usr/local/lib/python3.11/lib-dynload',  '/usr/local/lib/python3.11/site-packages']
> Fri, 04 Nov 2022 08:25:58 +0100
> ```

This is fixed in another branch

**Review by mattijsstam — APPROVED** [review]

**Review by mattijsstam — APPROVED** [review]

**Pull request merged**